pkg/controllers/certmanager/controller.go

```diff
@@ -2,91 +2,107 @@ package certmanager
 
 import (
 	"context"
-	"os"
-	"reflect"
 	"time"
 
+	"github.com/go-logr/logr"
 	"github.com/kyverno/kyverno/pkg/common"
 	"github.com/kyverno/kyverno/pkg/config"
 	"github.com/kyverno/kyverno/pkg/controllers"
 	"github.com/kyverno/kyverno/pkg/tls"
+	controllerutils "github.com/kyverno/kyverno/pkg/utils/controller"
 	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	corev1informers "k8s.io/client-go/informers/core/v1"
 	corev1listers "k8s.io/client-go/listers/core/v1"
-	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
 )
 
 const (
 	// Workers is the number of workers for this controller
 	Workers        = 1
 	ControllerName = "certmanager-controller"
+	maxRetries     = 10
 )
 
 type controller struct {
-	renewer      *tls.CertRenewer
+	renewer *tls.CertRenewer
+
+	// listers
 	secretLister corev1listers.SecretLister
-	secretQueue  chan bool
+
+	// queue
+	queue         workqueue.RateLimitingInterface
+	secretEnqueue controllerutils.EnqueueFunc
 }
 
 func NewController(secretInformer corev1informers.SecretInformer, certRenewer *tls.CertRenewer) controllers.Controller {
-	manager := &controller{
-		renewer:      certRenewer,
-		secretLister: secretInformer.Lister(),
-		secretQueue:  make(chan bool, 1),
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+	c := controller{
+		renewer:       certRenewer,
+		secretLister:  secretInformer.Lister(),
+		queue:         queue,
+		secretEnqueue: controllerutils.AddDefaultEventHandlers(logger.V(3), secretInformer.Informer(), queue),
 	}
-	secretInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    manager.addSecretFunc,
-		UpdateFunc: manager.updateSecretFunc,
+	return &c
+}
+
+func (c *controller) Run(ctx context.Context, workers int) {
+	go c.ticker(ctx)
+	// we need to enqueue our secrets in case they don't exist yet in the cluster
+	// this way we ensure the reconcile happens (hence renewal/creation)
+	c.secretEnqueue(&corev1.Secret{
```

Comment:
This is the interesting part, adding synthetic work in the queue so that it triggers reconcile.

Reply:
liked the idea !